File: Semestrul 4/MPP2/Problema8SC_CSharp/Persistence/RepoEvent.cs
using System.Data.SQLite;
using log4net;
using Persistence;
using Problema8SC_CSharp.Model;

namespace Problema8SC_CSharp.Persistence;

public class RepoEvent: IRepoEvent
{
    IDictionary<String, string> props;
    private SQLiteConnection _connection = DbUtils.GetConnection();
    private static readonly ILog log = LogManager.GetLogger("RepoEvent");

    public RepoEvent(IDictionary<String, string> props)
    {
        this.props = props;
    }
    
    public Event FindOne(long id)
    {
        log.Info($"Finding event {id}");
        String select = "SELECT * FROM event WHERE event_id = @Id";
        SQLiteCommand command = new SQLiteCommand(select, _connection);
        command.Parameters.AddWithValue("@Id", id);
        SQLiteDataReader reader = command.ExecuteReader();
        if (reader.Read())
        {
            var ev = new Event(
                (long)reader["event_id"],
                (string)reader["name"],
                (int)reader["distance"],
                (long)reader["age_group_id"]

            );
            log.Info($"Found event {ev}");
            return ev;
        }
        log.Info($"Event {id} not found");
        return null;
    }

    public IEnumerable<Event> FindAll()
    {
        log.Info("Finding all events");
        String select = "SELECT * FROM event";
        SQLiteCommand command = new SQLiteCommand(select, _connection);
        SQLiteDataReader reader = command.ExecuteReader();
        List<Event> events = new List<Event>();
        while (reader.Read())
        {
            var ev = new Event(
                (long)reader["event_id"],
                (string)reader["name"],
                (int)reader["distance"],
                (long)reader["age_group_id"]

            );
            events.Add(ev);
        }
        log.Info($"Found {events.Count} events");
        return events;
    }

    public void Save(Event entity)
    {
        log.Info($"Saving event {entity}");
        String insert = "INSERT INTO event (name, distance) VALUES (@Name, @Distance)";
        SQLiteCommand command = new SQLiteCommand(insert, _connection);
        command.Parameters.AddWithValue("@Name", entity.Name);
        command.Parameters.AddWithValue("@Distance", entity.Distance);
        command.ExecuteNonQuery();
    }

    public void Delete(long id)
    {
        log.Info($"Deleting event {id}");
        String delete = "DELETE FROM event WHERE event_id = @Id";
        SQLiteCommand command = new SQLiteCommand(delete, _connection);
        command.Parameters.AddWithValue("@Id", id);
        command.ExecuteNonQuery();
    }

    public void Update(Event entity)
    {
        log.Info($"Updating event {entity}");
        String update = "UPDATE event SET name = @Name, distance = @Distance WHERE event_id = @Id";
        SQLiteCommand command = new SQLiteCommand(update, _connection);
        command.Parameters.AddWithValue("@Name", entity.Name);
        command.Parameters.AddWithValue("@Distance", entity.Distance);
        command.Parameters.AddWithValue("@Id", entity.GetId());
        command.ExecuteNonQuery();
    }
    
    public IEnumerable<Event> GetEventsByAgeGroup(long ageGroupId)
    {
        log.Info($"Finding events for age group {ageGroupId}");
        String select = "SELECT * FROM event WHERE age_group_id = @AgeGroupId";
        SQLiteCommand command = new SQLiteCommand(select, _connection);
        command.Parameters.AddWithValue("@AgeGroupId", ageGroupId);
        SQLiteDataReader reader = command.ExecuteReader();
        List<Event> events = new List<Event>();
        while (reader.Read())
        {
            var ev = new Event(
                (long)reader["event_id"],
                (string)reader["name"],
                (int)reader["distance"],
                (long)reader["age_group_id"]

            );
            events.Add(ev);
        }
        log.Info($"Found {events.Count} events for age group {ageGroupId}");
        return events;
    }
}

File: Semestrul 4/MPP2/Problema8_SC_REST/Problema8SC_Rest/Problema8Persistence/src/main/java/problema8/persistence/jdbc/RepoEvent.java
package problema8.persistence.jdbc;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.springframework.stereotype.Component;
import org.springframework.stereotype.Repository;
import problema8.model.Event;
import problema8.persistence.RepositoryException;
import problema8.persistence.irepository.IRepoEvent;

import java.sql.*;
import java.util.ArrayList;
import java.util.Properties;

@Repository
public class RepoEvent implements IRepoEvent {
    private JdbcUtils jdbcUtils;

    private static final Logger logger = LogManager.getLogger();

    public RepoEvent(Properties properties) {
        logger.info("Initializing EventDBRepository with properties: {} ", properties);
        jdbcUtils = new JdbcUtils(properties);
    }

    @Override
    public Event findOne(Long aLong) {
        logger.info("Finding Event by id: {}", aLong);
        Connection connection = jdbcUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement("SELECT * FROM event WHERE event_id = ?")) {
            preparedStatement.setLong(1, aLong);
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                if (resultSet.next()) {
                    String name = resultSet.getString("name");
                    Integer distance = resultSet.getInt("distance");
                    Long id = resultSet.getLong("age_group_id");
                    Event event = new Event(aLong, name, distance, id);
                    logger.traceExit(event);
                    return event;
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
        }
        logger.traceExit("No Event found with id: {}", aLong);
        return null;
    }

    @Override
    public Iterable<Event> findAll() {
        logger.traceEntry("finding all Events");
        Connection connection = jdbcUtils.getConnection();
        ArrayList<Event> events = new ArrayList<>();
        try (PreparedStatement preparedStatement = connection.prepareStatement("SELECT * FROM event");
             ResultSet resultSet = preparedStatement.executeQuery();
        ) {
            while (resultSet.next()) {
                Long id = resultSet.getLong("event_id");
                String name = resultSet.getString("name");
                Integer distance = resultSet.getInt("distance");
                Long ag_id = resultSet.getLong("age_group_id");
                Event event = new Event(id, name, distance, ag_id);
                events.add(event);
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
        }
        logger.traceExit("No Events found");
        return events;
    }

    @Override
    public void save(Event entity) {
        logger.info("Saving Event: {}", entity);
        Connection connection = jdbcUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement("INSERT INTO event(name, distance, age_group_id) VALUES (?, ?)")) {
            preparedStatement.setString(1, entity.getName());
            preparedStatement.setInt(2, entity.getDistance());
            preparedStatement.setLong(3, entity.getAgeGroupId());
            preparedStatement.executeUpdate();
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
        }
    }

    public Event saveReturnEvent(Event entity) {
        logger.info("Saving Event and returning it: {}", entity);
        Connection connection = jdbcUtils.getConnection();

        try {
            String insertSql = "INSERT INTO event(name, distance, age_group_id) VALUES (?, ?, ?)";
            try (PreparedStatement insertStmt = connection.prepareStatement(insertSql)) {
                insertStmt.setString(1, entity.getName());
                insertStmt.setInt(2, entity.getDistance());
                insertStmt.setLong(3, entity.getAgeGroupId());
                insertStmt.executeUpdate();
            }

            try (Statement idStmt = connection.createStatement();
                 ResultSet rs = idStmt.executeQuery("SELECT last_insert_rowid()")) {
                if (rs.next()) {
                    long id = rs.getLong(1);
                    entity.setId(id);
                    logger.info("Inserted Event ID: {}", id);
                } else {
                    throw new SQLException("Failed to retrieve last insert ID.");
                }
            }

        } catch (SQLException e) {
            logger.error("Error saving Event", e);
            throw new RuntimeException("Database error: " + e.getMessage(), e);
        }

        return entity;
    }


    @Override
    public void delete(Long aLong) {
        logger.info("Deleting Event with id: {}", aLong);
        Connection connection = jdbcUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement("DELETE FROM event WHERE event_id = ?")) {
            preparedStatement.setLong(1, aLong);
            preparedStatement.executeUpdate();
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
        }
    }

    @Override
    public void update(Event entity) {
        logger.info("Updating Event: {}", entity);
        Connection connection = jdbcUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement("UPDATE event SET name = ?, distance = ?, age_group_id = ? WHERE event_id = ?")) {
            preparedStatement.setString(1, entity.getName());
            preparedStatement.setInt(2, entity.getDistance());
            preparedStatement.setLong(3, entity.getAgeGroupId());
            preparedStatement.setLong(4, entity.getId());
            preparedStatement.executeUpdate();
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
        }
    }

    public Iterable<Event> findByAgeGroup(Long ageGroupId) {
        logger.info("Finding Events by AgeGroup id: {}", ageGroupId);
        Connection connection = jdbcUtils.getConnection();
        ArrayList<Event> events = new ArrayList<>();
        try (PreparedStatement preparedStatement = connection.prepareStatement("SELECT * FROM event WHERE age_group_id = ?")) {
            preparedStatement.setLong(1, ageGroupId);
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    Long id = resultSet.getLong("event_id");
                    String name = resultSet.getString("name");
                    Integer distance = resultSet.getInt("distance");
                    Event event = new Event(id, name, distance, ageGroupId);
                    events.add(event);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
        }
        logger.traceExit(events);
        return events;
    }

    public void deleteByName(String name) {
        logger.info("Deleting Event by name: {}", name);
        Connection connection = jdbcUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement("DELETE FROM event WHERE name = ?")) {
            preparedStatement.setString(1, name);
            preparedStatement.executeUpdate();
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
        }
    }
}
